### DISCUSSION TITLE
Select First Row In Group

### DISCUSSION THREAD

**Discussion by MohMehrnia** [ravendb/ravendb#13605] [Q&A]
Hi,
I have a logs collection and I need to group by one column then select all columns in the first row of the group?

**Comment by omerlewitz** [comment]
Hi, 
Have you seen this - https://ravendb.net/docs/article-page/5.3/Csharp/client-api/session/querying/how-to-perform-group-by-query

**Comment by ayende** [comment]
Can you show the query that you would like to run?

**Reply by MohMehrnia** [reply]
`{
   "timestamp": ISODate("2015-09-27T15:28:06.0Z"),
    "value": '123'
},
{
   "timestamp": ISODate("2015-09-27T15:31:06.0Z"),
    "value": '737'
},
{
   "timestamp": ISODate("2015-09-27T15:35:00.0Z"),
    "value": '456'
},
{
   "timestamp": ISODate("2015-09-27T15:40:20.0Z"),
    "value": '789'
}`

Results Must Be:
`{
    "timestamp": ISODate("2015-09-27T15:25:00.0Z"),
    "value": '123'
},
{
    "timestamp": ISODate("2015-09-27T15:35:00.0Z"),
    "value": '456' // 456 has a newer timestamp than 737, which are in the same 5 minute range
},
{
    "timestamp": ISODate("2015-09-27T15:40:00.0Z"),
    "value": '789'
}`

**Reply by MohMehrnia** [reply]
In MongoDB Like this:
`db.WellRecords.aggregate([{
    $group: {
        "_id": { "$dateTrunc": { date: "$_id", unit: "minute", binSize: 15 } },
        "doc": { "$first": "$$ROOT" }
    }
},
{ "$replaceRoot": { "newRoot": "$doc" } },
{ "$sort": { "_id": 1 } }
])`

**Comment by ayende** [comment]
You can do that using an index, defined as:

```
from i in docs
let rounded = i.timestamp.AddMinutes(15 - i.timestamp.Minute % 15)
select new { 
    timestamp = rounded.ToString("yyyy-MM-ddTHH:mm"), 
    i.value 
    
}
```

```
from r in results
group r by r.timestamp into g
select new {
 timestamp = g.Key,
 value = g.Max(x=>x.value)
}
```

This results in:

![image](https://user-images.githubusercontent.com/116915/153191328-c63be9a2-8e06-4ee6-b743-d08dc8db11ad.png)

**Reply by MohMehrnia** [reply]
How can get all fields>

**Reply by ayende** [reply]
The map:

```
from i in docs
let rounded = i.timestamp.AddMinutes(15 - i.timestamp.Minute % 15)
select new { 
    timestamp = rounded.ToString("yyyy-MM-ddTHH:mm"), 
    value = new[]{ i }
   
}
```

The reduce:

```
from r in results
group r by r.timestamp into g
select new {
 timestamp = g.Key,
 values = g.OrderByDescending(x=>x.val).Take(1)
}
```
